### PR TITLE
Tests for new string.contains Api

### LINKED ISSUE #25094: Proposal: String.Contains(char)
## Rationale and Usage

Finding a character in a string is a fairly common primitive operation. Unfortunately we lead users to the pit of failure because `mystring.Contains(char)` will bind to the Linq version, which can easily be 20x slower than `mystring.IndexOf(char) != -1`. By adding these to string, the next recompile will give a performance improvement.

## Proposed API

```c#
    public sealed partial class String : System.Collections.Generic.IEnumerable<char>, System.Collections.IEnumerable, System.IComparable, System.IComparable<string>, System.IConvertible, System.IEquatable<string>, System.ICloneable
    {
        public bool Contains(char value) { throw null; }
        public bool Contains(char value, StringComparison comparisonType) { throw null; }
        public bool Contains(string value) { throw null; }  // already exists 
        public bool Contains(string value, StringComparison comparisonType) { throw null; } // already exists
        public int IndexOf(char value, StringComparison comparisonType) { throw null; } // to implement above
    }
```

the implementations will simply be

```c#
public bool Contains(char value)
{
    return IndexOf(value) != -1;
}

public bool Contains(char value, StringComparison comparisonType)
{
    return IndexOf(value, comparisonType) != -1;
}
```

## Microbenchmark

Searching in 10 and 1000 character strings:
```
    Method |       Mean |     Error |    StdDev |
---------- |-----------:|----------:|----------:|
 LinqShort | 142.750 ns | 0.4997 ns | 0.4430 ns |
  LinqLong | 180.121 ns | 0.4741 ns | 0.3428 ns |
 FastShort |   7.575 ns | 0.0196 ns | 0.0174 ns |
  FastLong |   8.508 ns | 0.0403 ns | 0.0377 ns |
```

## Variations

The StringComparison overload is in order to search case insensitively if desired. The IndexOf overload is needed to implement it. These could be discarded as the 90% case I would expect to not take a comparison.

### PR DESCRIPTION
**Pull request by Anipik**
Fixes #25094
PR implementation link = https://github.com/dotnet/coreclr/pull/15180

### CODE REVIEW

**Comment by danmosemsft** [comment]
Please use a better title eg "Tests for new String.Contains API"
Also include a link to the PR for the implementation

**Comment by danmosemsft** [comment]
Looks good but needs tests for the other 2 api

File: src/System.Runtime/tests/System/StringTests.cs [hunk 1]
```diff
@@ -836,6 +836,59 @@ public static void Contains(string s, string value, bool expected)
...
+        // OrdinalIgnoreCase
+        [InlineData("Hello", 'H', StringComparison.OrdinalIgnoreCase, true)]
+        [InlineData("Hello", 'Z', StringComparison.OrdinalIgnoreCase, false)]
+        [InlineData("Hello", 'e', StringComparison.OrdinalIgnoreCase, true)]
+        [InlineData("Hello", 'E', StringComparison.OrdinalIgnoreCase, true)]
+        [InlineData("", 'H', StringComparison.OrdinalIgnoreCase, false)]
+        public static void Contains(string s, char value, StringComparison comparisionType, bool expected)
+        {
+            Assert.Equal(expected, s.Contains(value, comparisionType));
+        }
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.cs]: These tests are for new APIs and will fail to compile here when building against netstandard.  They should be moved to StringTests.netcoreapp.cs.

File: src/System.Runtime/ref/System.Runtime.cs [hunk 2]
```diff
@@ -2190,7 +2190,9 @@ public sealed partial class String : System.Collections.Generic.IEnumerable<char
         public static string Concat(params string[] values) { throw null; }
         public static string Concat<T>(System.Collections.Generic.IEnumerable<T> values) { throw null; }
         public bool Contains(string value) { throw null; }
+        public bool Contains(char value) { throw null; }
```
Comment (danmosemsft) [inline src/System.Runtime/ref/System.Runtime.cs]: Nit: please sort : c comes before s.

**Comment by Anipik** [comment]
@dotnet-bot test this please

**Comment by danmosemsft** [comment]
@Anipik the UWP builds is failing because the API is not ported into CoreRt. Please add the implementation in the CoreRt repo like you did in coreclr.

**Comment by Anipik** [comment]
@danmosemsft 
https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/shared/System/String.Searching.cs
It already contains the implementation

**Comment by danmosemsft** [comment]
Ah..its shared, you're right. The issue then is it needs to mirror again into the project N branch which might take a few days. You can just sit on this PR and wait for that.

**Comment by Anipik** [comment]
@dotnet-bot test this please

File: src/System.Runtime/ref/System.Runtime.cs [hunk 3]
```diff
@@ -2215,6 +2217,7 @@ public sealed partial class String : System.Collections.Generic.IEnumerable<char
         public override int GetHashCode() { throw null; }
         public int GetHashCode(System.StringComparison comparisonType) { throw null; }
         public int IndexOf(char value) { throw null; }
+        public int IndexOf(char value, StringComparison comparisonType) { throw null; }
```
Comment (eerhardt) [inline src/System.Runtime/ref/System.Runtime.cs]: Are there tests for this new `IndexOf` overload?

**Comment by danmosemsft** [comment]
@Anipik as a general rule please wait for at least one "green checkmark" signoff before merging. (My "Looks good but needs tests for the other 2 api" may have been confusing)

File: src/System.Runtime/ref/System.Runtime.cs (hunk 3 above)
Comment (Anipik) [reply src/System.Runtime/ref/System.Runtime.cs]: yah

Comment (eerhardt) [reply src/System.Runtime/ref/System.Runtime.cs]: I missed them. Can you point me to them?

Comment (Anipik) [reply src/System.Runtime/ref/System.Runtime.cs]: src/System.Runtime/tests/System/StringTests.netcoreapp.cs

Comment (eerhardt) [reply src/System.Runtime/ref/System.Runtime.cs]: I’m sorry I can’t seem to find them. I must be missing something. Can you provide a direct link?  I see the new Contains tests you added, but no tests for the new IndexOf overload.

Comment (Anipik) [reply src/System.Runtime/ref/System.Runtime.cs]: https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/StringTests.netcoreapp.cs#L141

Comment (stephentoub) [reply src/System.Runtime/ref/System.Runtime.cs]: @Anipik, that's a test of Contains, not IndexOf.  @eerhardt is asking about the new IndexOf overload, for which no new tests were added.

Comment (Anipik) [reply src/System.Runtime/ref/System.Runtime.cs]: Sorry you are right. I am not directly testing indexof. I will add tests

Comment (Anipik) [reply src/System.Runtime/ref/System.Runtime.cs]: @stephentoub  should they be added  to stringtests.cs or stringtests.netcoreapp.cs?

Comment (stephentoub) [reply src/System.Runtime/ref/System.Runtime.cs]: > should they be added to stringtests.cs or stringtests.netcoreapp.cs?

They won't compile if they're added to StringTests.cs.  They should be added to StringTests.netcoreapp.cs.

**Pull request merged**